Main.hs

```diff
@@ -2,37 +2,115 @@ module Main where
 
 import Control.Monad (unless)
 import Text.Printf (printf)
+import Distribution.Compat.ResponseFile (expandResponse)
+import Text.XHtml (base, abbr, reset)
+
+data Expr = Num Double
+          | Sqrt Expr
+          | Add Expr Expr
+          | Sub Expr Expr
+          | Mul Expr Expr
+          | Div Expr Expr
+          | Pow Expr Expr
 
-data Expr
 
 instance Show Expr where
-  show = undefined
+  show (Num x) = show x
+  show (Sqrt expr) = "sqrt(" ++ show expr ++ ")"
+  show (Add expr1 expr2) = "(" ++ show expr1 ++ ") + (" ++ show expr2 ++ ")"
+  show (Sub expr1 expr2) = "(" ++ show expr1 ++ ") - (" ++ show expr2 ++ ")"
+  show (Mul expr1 expr2) = "(" ++ show expr1 ++ ") * (" ++ show expr2 ++ ")"
+  show (Div expr1 expr2) = "(" ++ show expr1 ++ ") / (" ++ show expr2 ++ ")"
+  show (Pow expr1 expr2) = "(" ++ show expr1 ++ ") ^ (" ++ show expr2 ++ ")"
 
 instance Eq Expr where
-  (==) = undefined
+  a == b = show a == show b
+
 
-data Error
+data Error = ZeroDivision
+           | NegativeNumberSqrt
 
 instance Show Error where
-  show = undefined
+  show :: Error -> String
+  show ZeroDivision = "Division by zero error"
+  show NegativeNumberSqrt = "Taking square root from a negative number error"
+
 
 instance Eq Error where
-  (==) = undefined
+  ZeroDivision == ZeroDivision = True
+  NegativeNumberSqrt == NegativeNumberSqrt = True
+  _ == _ = False
```

Comment:
Whenever your implementation of a type class is trivial, it's better to derive it.